README.md
## nora-wos

Harvest and process Web of Science (WoS) and Incites data.

### Requirements

To build the debian package you will need the following package:
* make
* devscripts
* debhelper

You can build the package by running:
* perl Makefile.PL
* make debian

### Running nora-wos (data backend)

Backup database before update:
* `$ nora-wos system backup`

Update WoS data (on the 19th of each month):
* `$ nora-wos update danish-records`
     
Generate DOI/orgs file (after WoS update):
* `$ nora-wos generate orgs-doi`

Update Incites indicators (end of month, see https://incites.help.clarivate.com/Content/dataset-updates.htm)
* `$ nora-wos nora-wos fetch indicators`

### Running nora-wos (VIVO side)

Generate RDF (after Incites update):
* `$ nora-wos rdf generate`


debian/changelog
nora-wos (0.04-00) unstable; urgency=low

  * Fixes to RDF gen
    
    GIT revision: ccd2cc237085f706b2dde28b6ccfabb98473241a

 -- kybus <fkybus@gmail.com>  Mon, 28 Nov 2022 14:29:24 +0100

nora-wos (0.03-00) unstable; urgency=low

  * A start of split between Data management and RDF processing
    
    GIT revision: ea5350a52e2fec127f8a599ec4c691531708e44e

 -- kybus <fkybus@gmail.com>  Thu, 22 Sep 2022 10:59:14 +0200

nora-wos (0.02-00) unstable; urgency=low

  * Commited all changes so far, we be use for current update and then
    updated to pass it on to Nelson and Brian.
    
    GIT revision: 03e3111f0034ab024a3756770cffedb5dbe30842

 -- kybus <fkybus@gmail.com>  Sat, 10 Sep 2022 19:39:56 +0200

nora-wos (0.01-00) unstable; urgency=low

  * Development version

 -- franck <fkybus@gmail.com>  Thu, 25 Nov 2021 10:34:16 +0100



Makefile.PL
use 5.020002;
use ExtUtils::MakeMaker;
WriteMakefile(
    NAME              => 'NORA::WOS',
    VERSION_FROM      => 'lib/NORA/WOS.pm',
    PREREQ_PM         => {},
    AUTHOR            => 'franck <fkybus@gmail.com>',
);

sub MY::postamble
{
    return <<'MAKE_FRAG';
debian: all
	@dpkg-buildpackage -rfakeroot -uc -us

install::
	@mkdir -p $(DESTDIR)/etc/nora-wos
	@mkdir -p $(DESTDIR)/var/log/nora-wos
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/db
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/cache
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/data
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/rdf
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/staging
	@mkdir -p $(DESTDIR)/var/lib/nora-wos/vivo
	@mkdir -p $(DESTDIR)/var/run/nora-wos
	@install -D -m 775 bin/nora-wos $(DESTDIR)/usr/bin/nora-wos
	@for file in sbin/*; \
        do \
            install -D -m 775 $$file $(DESTDIR)/usr/$$file; \
        done
	@for file in etc/*; \
        do \
            ffile=`basename $$file`; \
            install -D -m 664 $$file $(DESTDIR)/etc/nora-wos/$$ffile; \
        done

distclean::
	rm -rf debian/nora-wos
	rm -f debian/files debian/nora-wos.substvars build-stamp install-stamp
MAKE_FRAG
}



lib/NORA/WOS.pm
package NORA::WOS;

use strict;
use warnings;
use Digest::MD5 qw(md5_hex);

our $VERSION = '0.01';

sub new
{
    my ($class) = @_;

    my $self = {};
    if (open (my $fin, '/etc/nora-wos/conf.tab')) {
        while (<$fin>) {
            chomp;
            my ($key, $val) = split (' ', $_, 2);
            $self->{'conf'}{$key} = $val;
        }
        close ($fin);
    } else {
        die ("fatal: failed to open /etc/nora-wos/conf.tab for reading: $!\n");
    }
    $self->{'field-variant'} = {
        language => [
            'static_data/fullrecord_metadata/normalized_languages/language[type=primary]/content',
            'static_data/fullrecord_metadata/languages/language[type=primary]/content',
            'static_data/fullrecord_metadata/normalized_languages[]/language/content',
            'static_data/fullrecord_metadata/languages/language[]/content',
        ],
    };
    return (bless ($self, $class));
}

sub conf
{
    my ($self, $key) = @_;

    return ($self->{'conf'}{$key});
}

sub stat_start
{
    my ($self, $name) = @_;

    $self->{'stat'}{$name} = time;
}

sub stat_time
{
    my ($self, $name) = @_;

    return (time - $self->{'stat'}{$name});
}

sub stat_display_time
{
    my ($self, $name) = @_;

    return ($self->display_time (time - $self->{'stat'}{$name}));
}

sub display_time
{
    my ($self, $sec) = @_;

    my $hour = int ($sec / 3600);
    $sec = ($sec % 3600);
    my $min = int ($sec / 60);
    $sec = ($sec % 60);
    if ($hour) {
        return (sprintf ('%d:%2d:%02d', $hour, $min, $sec));
    } else {
        return (sprintf ('%2d:%02d', $min, $sec));
    }
}

sub counter
{
    my ($self, $name, $inc) = @_;

    if (!exists ($self->{'counter'}{$name})) {
        $self->{'counter'}{$name} = 0;
    }
    if ($inc) {
        if ($inc eq 'reset') {
            $self->{'counter'}{$name} = 0;
        } else {
            $self->{'counter'}{$name} += $inc;
        }
    }
    return ($self->{'counter'}{$name});
}

sub eta
{
    my ($self, $time) = @_;

    $time += time;
    my ($sec, $min, $hour, $day, $mon, $year) = localtime (int ($time));
    return (sprintf ('%04d-%02d-%02d %02d:%02d:%02d', 1900 + $year, $mon + 1, $day, $hour, $min, $sec));
}

sub date
{
    my ($self, $time) = @_;

    if (!$time) {
        $time = time;
    }
    my ($sec, $min, $hour, $day, $mon, $year) = localtime ($time);
    return (sprintf ('%04d-%02d-%02d', 1900 + $year, $mon + 1, $day));
}

sub search_key
{
    my ($self, @fld) = @_;

    my $words = {};
    foreach my $f (@fld) {
        $f =~ s/[^0-9A-Za-z\.]+/ /g;
        $f =~ s/([0-9])\.([0-9])/$1-$2/g;
        $f =~ s/[\.\s]+/ /g;
        $f =~ s/-/./g;
        $f =~ s/^\s//;
        $f =~ s/\s$//;
        foreach my $w (split (' ', lc ($f))) {
            $words->{$w} = 1;
        }
    }
    return (sort (keys (%{$words})));
}

sub field
{
    my ($self, $id, $rec, $path, $filter) = @_;

    if ($filter) {
        if (ref ($rec) ne 'HASH') {
            $self->log ('e', '%s: not a HASH for filter %s: %s', $id, $filter, $path);
            return (undef);
        }
        my ($fld, $val) = split ('=', $filter);
        if (!exists ($rec->{$fld})) {
            return (undef);
        }
        if (($val) && ($rec->{$fld} ne $val)) {
            return (undef);
        }
    }
    my @path = split ('/', $path);
    my $e;
    while ($e = shift (@path)) {
        if (ref ($rec) ne 'HASH') {
            $self->log ('e', '%s: not a HASH for %s: %s', $id, $e, $path);
            return (undef);
        }
        if ($e =~ s/\]$//) {
            my @val = ();
            my ($fld, $flt) = split (/\[/, $e);
            if (!exists ($rec->{$fld})) {
                return ();
            }
            if (ref ($rec->{$fld}) eq 'HASH') {
                $rec->{$fld} = [$rec->{$fld}];
            }
            if (ref ($rec->{$fld}) eq 'ARRAY') {
                foreach my $rc (@{$rec->{$fld}}) {
                    if (($flt) || (@path)) {
                        my @res = $self->field ($id, $rc, join ('/', @path), $flt);
                        if (@res) {
                            push (@val, @res);
                        }
                    } else {
                        if ($rc) {
                            push (@val, $rc);
                        }
                    }
                }
            } else {
                if ((ref ($rec->{$fld}) eq '') && (!$flt)) {
                    push (@val, $rec->{$fld});
                } else {
                    $self->log ('e', '%s: not an ARRAY for %s: %s - %s', $id, $fld, $path, ref ($rec->{$fld}));
                    return (undef);
                }
            }
            return (@val);
        } else {
            if (exists ($rec->{$e})) {
                $rec = $rec->{$e}
            } else {
                return ();
            }
        }
    }
    return ($rec);
}

sub field_map
{
    my ($self, $name, $key, $value) = @_;

    if (!defined ($key)) {
        $self->{'field_map'}{$name} = {};
        return ($self->{'field_map'}{$name});
    }
    if (!exists ($self->{'field_map'}{$name})) {
        $self->log ('w', 'using undefined file_map: %s', $key);
        $self->{'field_map'}{$name} = {};
    }
    if (defined ($name)) {
        $self->{'field_map'}{$name}{$key} = $name;
    }
    return ($self->{'field_map'}{$name}{$key});
}

sub key_id
{
    my ($self, $name) = @_;

    $name = lc ($name);
    $name =~ s/[^[:alnum:]]+/ /g;
    $name =~ s/^\s//;
    $name =~ s/\s$//;
    my $md5 = $name;
    utf8::encode($md5);
    return ($name, md5_hex ($md5));
}

sub field_variant
{
    my ($self, $id, $rec, $name) = @_;

    if (!exists ($self->{'field-variant'}{$name})) {
        $self->log ('f', '%s: field variant not defined: %s', $id, $name);
        exit (1);
    }
    my @val = ();
    foreach my $path (@{$self->{'field-variant'}{$name}}) {
        if (@val = $self->field ($id, $rec, $path)) {
#           $self->log ('d', '"%s" got: "%s"', $path, join (';; ', @val));
            return (@val);
        }
#       $self->log ('d', '"%s" got: null', $path);
    }
    return (undef);
}

sub country_code
{
    my ($self, $name) = @_;

    if (!exists ($self->{'country-code'})) {
        if (!-e '/etc/nora-wos/countries.tsv') {
            $self->log ('f', 'missing country mapping: /etc/nora-wos/countries.tsv');
            exit (1);
        }
        if (open (my $fin, '/etc/nora-wos/countries.tsv')) {
            $self->{'country-code'} = {};
            while (<$fin>) {
                chomp;
                my ($key, $code) = split ("\t");
                $self->{'country-code'}{$key} = $code;
            }
            close ($fin);
        } else {
            $self->log ('f', 'failed to open /etc/nora-wos/countries.tsv for reading: %s', $!);
            exit (1);
        }
    }
    my ($key) = $self->key_id ($name);
    if ($self->{'country-code'}{$key}) {
        return ($self->{'country-code'}{$key});
    } else {
        $self->log ('w', 'missing country mapping for "%s" (%s)', $name, $key);
        return ('');
    }
}

sub log
{
    my ($self, $level, $msg, @args) = @_;

    my ($sec, $min, $hour, $day, $mon, $year) = localtime (time);
    printf (STDERR "%04d-%02d-%02d %02d:%02d:%02d %s $msg\n", 1900 + $year, $mon + 1, $day, $hour, $min, $sec, $level, @args);
}

1;



lib/NORA/WOS/DB.pm
package NORA::WOS::DB;

use strict;
use warnings;
use NORA::WOS;
use JSON::XS;
use DBI qw(:sql_types);

sub new
{
    my ($class, $dbfile) = @_;
    my $self = {};

    if (-e '/var/lib/nora-wos/db') {
        $self->{'dbfile'} = '/var/lib/nora-wos/db/nora-wos.sqlite3';
    } else {
        $self->{'dbfile'} = 'nora-wos.sqlite3';
    }
    $self->{'dbh'} = DBI->connect ('dbi:SQLite:' . $self->{'dbfile'}, undef, undef, {PrintError => 1, RaiseError => 0});
    $self->{'wos'} = new NORA::WOS ();
    return (bless ($self, $class));
}

sub reconnect
{
    my ($self) = @_;

    $self->{'dbh'} = DBI->connect ('dbi:SQLite:' . $self->{'dbfile'}, undef, undef, {PrintError => 1, RaiseError => 0});
}

sub dbfile
{
    my ($self) = @_;

    return ($self->{'dbfile'});
}

sub doctype
{
    my $type = {
        article => 1,
        review  => 1,
    };
}

sub ind_update
{
    my ($self, $rec) = @_;

    if ($rec->{'stamp'}) {
        $rec->{'stamp'} = time;
        my @fld = ();
        my @val = ();
        foreach my $key (sort (keys (%{$rec}))) {
            if (exists ($self->{'ind'}{$key})) {
                push (@fld, $self->{'ind'}{$key} . '=?');
                push (@val, $rec->{$key});
            } else {
                $self->log ('f', 'unmapped Incites indicator: "%s"', $key);
                exit (1);
            }
        }
        $self->sql ('update ind set ' . join (',', @fld) . " where ut='" . $rec->{'ut'} . "'");
    } else {
        $rec->{'stamp'} = time;
        my @fld = ();
        my @pla = ();
        my @val = ();
        foreach my $key (sort (keys (%{$rec}))) {
            if (exists ($self->{'ind'}{$key})) {
                push (@fld, $self->{'ind'}{$key});
                push (@pla, '?');
                push (@val, $rec->{$key});
            } else {
                $self->log ('f', 'unmapped Incites indicator: "%s"', $key);
                exit (1);
            }
        }
        $self->sql ('insert into ind (' . join (',', @fld) . ') VALUES (' . join (',', @pla) . ')', @val);
    }
}

sub doc_core_update
{
    my ($self, $rec) = @_;

    if (($rec->{'stamp'}) || ($self->{'doc_core_update'}{$rec->{'ut'}})) {
        if (!$rec->{'stamp'}) {
            warn ("info: update double: $rec->{'ut'}\n");
        }
        $rec->{'stamp'} = time;
        my @fld = ();
        my @val = ();
        foreach my $key (sort (keys (%{$rec}))) {
            push (@fld, $key . '=?');
            push (@val, $rec->{$key});
        }
        $self->sql ('update doc_core set ' . join (',', @fld) . " where ut='" . $rec->{'ut'} . "'");
    } else {
        $self->{'doc_core_update'}{$rec->{'ut'}} = 1;
        $rec->{'stamp'} = time;
        my @fld = ();
        my @pla = ();
        my @val = ();
        foreach my $key (sort (keys (%{$rec}))) {
            push (@fld, $key);
            push (@pla, '?');
            push (@val, $rec->{$key});
        }
        $self->sql ('insert into doc_core (' . join (',', @fld) . ') VALUES (' . join (',', @pla) . ')', @val);
    }
}

sub doc_person_insert
{
    my ($self, $rec) = @_;

    $rec->{'stamp'} = time;
    my @fld = ();
    my @pla = ();
    my @val = ();
    foreach my $key (sort (keys (%{$rec}))) {
        push (@fld, $key);
        push (@pla, '?');
        push (@val, $rec->{$key});
    }
    $self->sql ('insert into doc_person (' . join (',', @fld) . ') VALUES (' . join (',', @pla) . ')', @val);
}

sub close
{
    my ($self) = @_;

    $self->{'dbh'}->disconnect ();
}

sub json_load
{
    my ($file) = @_;

#   open (my $fin, $file);
#   my $json = join ('', <$fin>);
#   close ($fin);
#   return (decode_json ($json));
}

sub create
{   
    my ($self) = @_;

    $self->sql ('create table if not exists doc (
                     ut                  string primary key,
                     stamp               integer,
                     year                integer,
                     doctype             string,
                     title               string,
                     authors             string,
                     source              string,
                     volume              string,
                     issue               string,
                     pubdate             string,
                     cited               integer,
                     refs                integer,
                     doi                 string,
                     json                string)'); 
    $self->sql ('create table if not exists indicator (
                     ut                  string primary key,
                     stamp               integer,
                     incites_date        string,
                     json                string)'); 
    $self->sql ('create index if not exists indicator_incites_date on indicator (incites_date)');
    $self->sql ('create table if not exists updates (
                     id                  integer primary key autoincrement,
                     stamp               integer,
                     upd                 string,
                     updlong             string)'); 
}

sub ind_mapping
{
    my ($self, $key) = @_;

    if (defined ($key)) {
        if (exists ($self->{'ind'}{$key})) {
            return ($self->{'ind'}{$key});
        } else {
            die ("fatal: unmapped indicator: '$key'");
        }
    } else {
        return ($self->{'ind'});
    }
}

sub update
{
    my ($self, $table, $key, $rec) = @_;

    if (($key) && ($rec->{$key})) {
        $rec->{'stamp'} = time;
        my @fld = ();
        my @val = ();
        foreach my $fld (sort (keys (%{$rec}))) {
            push (@fld, $fld . '=?');
            push (@val, $self->trim ($rec->{$fld}));
        }
        if ($rec->{$key} =~ m/^[0-9]+$/) {
            $self->sql ("update $table set " . join (',', @fld) . " where $key=" . $rec->{$key}, @val);
        } else {
            $self->sql ("update $table set " . join (',', @fld) . " where $key='" . $rec->{$key} . "'", @val);
        }
    } else {
        $rec->{'stamp'} = time;
        my @fld = ();
        my @pla = ();
        my @val = ();
        foreach my $fld (sort (keys (%{$rec}))) {
            push (@fld, $fld);
            push (@pla, '?');
            push (@val, $self->trim ($rec->{$fld}));
        }
        $self->sql ("insert into $table (" . join (',', @fld) . ') VALUES (' . join (',', @pla) . ')', @val);
    }
}

sub trim
{
    my ($self, $txt) = @_;

    if (!defined ($txt)) {
        return ('');
    }
    $txt =~ s/^[\s\t\r\n]+//;
    $txt =~ s/[\s\t\r\n]+$//;
    return ($txt);
}

sub sql
{
    my ($self, $sql, @args) = @_;

#   print ("debug: SQL: $sql\n");
    my $sth = $self->prepare ($sql);
    $self->execute ($sth, @args);
    return ($sth);
}

sub prepare
{
    my ($self, $sql) = @_;
    my ($sth);

    my $dbh = $self->{'dbh'};
    if ($sth = $dbh->prepare ($sql)) {
        return ($sth);
    } else {
        die ('fatal: ' . $dbh->errstr . " ($sql)");
    }
}

sub execute
{
    my ($self, $sth, @args) = @_;
    my ($status);

    $sth->execute (@args);
    if ((defined ($sth->errstr)) && (length ($sth->errstr))) {
        die ('fatal: ' . $sth->errstr);
    } else {
        return (1);
    }
}

sub ct_micro
{
    my ($self, $ut) = @_;

    if (!exists ($self->{'ct_micro'})) {
        my $oamap = {
            'free to read'    => 'publisherfree2read',
            'gold'            => 'publisherfullgold',
            'gold hybrid'     => 'publisherhybridgold',
            'green accepted'  => 'repository',
            'green only'      => 'repository',
            'green published' => 'repository',
            'green submitted' => 'repository',
        };
        $self->{'wos'}->log ('i', 'loading indicator micro CT');
        my $jxs = JSON::XS->new->allow_nonref->canonical(1);
        my $rs = $self->sql ('select ut,json from indicator');
        my $rc;
        while ($rc = $rs->fetchrow_hashref) {
            my $ind = $jxs->decode ($rc->{'json'});
            foreach my $cat (@{$ind->{'PERCENTILE'}}) {
                if ($cat->{'LEVEL'} == 3) {
                    my $ct = $cat->{'SUBJECT'};
                    $ct =~ s/^\s+//;
                    $ct =~ s/\s+.*//;
                    $ct =~ s/[^0-9\.]//g;
                    if ($ct) {
                        if (exists ($self->{'ct_micro'}{$rc->{'ut'}})) {
                            push (@{$self->{'ct_micro'}{$rc->{'ut'}}}, $ct);
                        } else {
                            $self->{'ct_micro'}{$rc->{'ut'}} = [$ct];
                        }
                    }
                }
            }
            if ($ind->{'OPEN_ACCESS'}{'OA_FLAG'}) {
                my $types = {all => 1};
                foreach my $status (@{$ind->{'OPEN_ACCESS'}{'STATUS'}}) {
                    my $type = lc ($status->{'TYPE'});
                    $type =~ s/[^a-z]+/ /g;
                    $type =~ s/^ //;
                    $type =~ s/ $//;
                    if ($oamap->{$type}) {
                        $types->{$oamap->{$type}} = 1;
                    } else {
                        $self->{'wos'}->log ('w', 'unknown OA type: %s (%s)', $type, $status->{'TYPE'});
                    }
                }
                $self->{'open_access'}{$rc->{'ut'}} = join (' ', sort (keys (%{$types})));
            }
        }
        $self->{'wos'}->log ('i', '- done');
    }
    if ($ut) {
        if (exists ($self->{'ct_micro'}{$ut})) {
            return (@{$self->{'ct_micro'}{$ut}});
        } else {
            return ();
        }
    } else {
        return ();
    }
}

sub open_access
{
    my ($self, $ut) = @_;

    if (!exists ($self->{'open_access'})) {
        $self->ct_micro ($ut);
    }
    if ($ut) {
        if (exists ($self->{'open_access'}{$ut})) {
            return ($self->{'open_access'}{$ut});
        } else {
            return ();
        }
    } else {
        return ();
    }
}

1;



lib/NORA/WOS/DOI.pm
package NORA::WOS::DOI;

use strict;
use warnings;
use NORA::WOS;
use NORA::WOS::DB;
use LWP::UserAgent;
use HTTP::Request;
use IO::Socket::SSL qw(SSL_VERIFY_NONE);

sub new
{
    my ($class) = @_;
    my $self = {baseurl => 'https://api.operanora.qa.dataverz.com/search/'};
    $self->{'baseurl'} = 'https://65.21.141.182/search/';
    $self->{'wos'} = new NORA::WOS ();
    $self->{'db'} = new NORA::WOS::DB ();
    $self->{'db'}->create ();

    return (bless ($self, $class));
}

sub normalize
{
    my ($self, $doi) = @_;

    if (!$doi) {
        return ();
    }
    $doi =~ s/[\s\t\r\n]//g;
    $doi =~ s/^.*?10\./10./;
    $doi =~ s/[\.,]+$//;
    if ($doi =~ m/^10\./) {
        return (lc ($doi));
    }
    return ();
}

sub harvest
{
    my ($self, $doi) = @_;

    my $url = $self->{'baseurl'} . '?id=' . $doi;
    my $rs = $self->http_request ($url);
    if ($rs->code != 200) {
        $self->{'wos'}->log ('e', 'DOI service error "%s", %s - %s', $url, $rs->code, $rs->message);
        $self->{'wos'}->log ('e', 'DOI service error "%s", %s', $url, $rs->status_line());
        return (undef);
    } else {
        return ($rs->content);
    }
}

sub harvest_batch
{
    my ($self) = @_;

    if (!exists ($self->{'skip'})) {
        $self->{'skip'} = 0;
    }
    my $url = $self->{'baseurl'} . '?limit=1000&skip=' . $self->{'skip'};
    $self->{'skip'} += 1000;
    my $rs = $self->http_request ($url);
    if ($rs->code != 200) {
        $self->{'wos'}->log ('e', 'DOI service error "%s", %s - %s', $url, $rs->code, $rs->message);
        return (undef);
    } else {
        return ($rs->content);
    }
}

sub http_request
{
    my ($self, $url) = @_;

    my $ua;
    if (exists ($self->{'ua'})) {
        $ua = $self->{'ua'};
    } else {
        $ua = new LWP::UserAgent;
        $ua->agent ('NORA-WOS-DOI/0.1');
        $ua->timeout (180);
        $ua->ssl_opts (verify_hostname => 0, SSL_verify_mode => 0x00);
        $self->{'ua'} = $ua;
    }
    my $re = new HTTP::Request (GET => $url);
    $re->header ('Accept' => 'application/json');
    my $rs = $ua->request ($re);
    return ($rs);
}

sub lookup
{
    my ($self, $doi) = @_;

    if (!($doi = $self->normalize ($doi))) {
        return ();
    }
    if (!exists ($self->{'lookup'})) {
        my $rs = $self->{'db'}->sql ('select doi,bfiLevel,bfiMra,bfiPubYear,bfiSubYear,oaiClass,oaiMra,oaiPubYear,oaiSubyear,dimID from doi');
        my $rc;
        while ($rc = $rs->fetchrow_hashref) {
            $self->{'lookup'}{$rc->{'doi'}} = $rc;
        }
    }
    return ($self->{'lookup'}{$doi});
}

1;



lib/NORA/WOS/N3.pm
package NORA::WOS::N3;

use strict;
use warnings;

our $VERSION = '0.01';

sub new
{
    my ($class) = @_;

    my $self = {};
    $self->{'prefix'} = {
        'bibo'         => 'http://purl.org/ontology/bibo/',
        'c4o'          => 'http://purl.org/spar/c4o/',
        'cito'         => 'http://purl.org/spar/cito/',
        'dcterms'      => 'http://purl.org/dc/terms/',
        'event'        => 'http://purl.org/NET/c4dm/event.owl#',
        'fabio'        => 'http://purl.org/spar/fabio/',
        'foaf'         => 'http://xmlns.com/foaf/0.1/',
        'geo'          => 'http://aims.fao.org/aos/geopolitical.owl#',
        'norao'        => 'https://www.nora.dtu.dk/ontology/nora/',
        'obo'          => 'http://purl.obolibrary.org/obo/',
        'ocrer'        => 'http://purl.org/net/OCRe/research.owl#',
        'ocresd'       => 'http://purl.org/net/OCRe/study_design.owl#',
        'ocresp'       => 'http://purl.org/net/OCRe/study_protocol.owl#',
        'ocresst'      => 'http://purl.org/net/OCRe/statistics.owl#',
        'osrap'        => 'http://vivo.deffopera.dk/ontology/osrap/',
        'owl'          => 'http://www.w3.org/2002/07/owl#',
        'rdf'          => 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
        'rdfs'         => 'http://www.w3.org/2000/01/rdf-schema#',
        'ro'           => 'http://purl.obolibrary.org/obo/ro.owl#',
        'scires'       => 'http://vivoweb.org/ontology/scientific-research#',
        'skos'         => 'http://www.w3.org/2004/02/skos/core#',
        'swo'          => 'http://www.ebi.ac.uk/efo/swo/',
        'swrl'         => 'http://www.w3.org/2003/11/swrl#',
        'swrlb'        => 'http://www.w3.org/2003/11/swrlb#',
        'vann'         => 'http://purl.org/vocab/vann/',
        'vcard'        => 'http://www.w3.org/2006/vcard/ns#',
        'vitro'        => 'http://vitro.mannlib.cornell.edu/ns/vitro/0.7#',
        'vitro-public' => 'http://vitro.mannlib.cornell.edu/ns/vitro/public#',
        'vivo'         => 'http://vivoweb.org/ontology/core#',
        'wos'          => 'https://wos.nora.dtu.dk/individual/',
        'xsd'          => 'http://www.w3.org/2001/XMLSchema#',
        'opera'        => 'http://vivo.deffopera.dk/ontology/osrap/',
        'orcid'        => 'https://orcid.org/',
        'cla'          => 'https://www.nora.dtu.dk/ontology/clarivate/',
    };
    $self->{'pused'} = {};
    $self->{'data'} = {};
    return (bless ($self, $class));
}

sub prefix
{
    my ($self, $code, $uri) = @_;

    if ($code !~ m/^[-0-9A-Za-z]+$/) {
        die ("fatal: invalid character in prefix code '$code', valid characters are: 0-9, A-Z, a-z and hyphen (-).\n");
    }
    $self->{'prefix'}{$code} = $uri;
}

sub add
{
    my ($self, $type, $id, $code, @val) = @_;

    $id = $self->pused ('id', $id);
    $code = $self->pused ('code', $code);
    if (!exists ($self->{'data'}{$type}{$id})) {
        $self->{'data'}{$type}{$id} = {};
    }
    if (!exists ($self->{'data'}{$type}{$id}{$code})) {
        $self->{'data'}{$type}{$id}{$code} = {};
    }
    foreach my $val (@val) {
        $val = $self->pused ('value', $val);
        $self->{'data'}{$type}{$id}{$code}{$val} = 1;

    }
}

sub add_text
{
    my ($self, $type, $id, $code, @val) = @_;

    my @VAL = ();
    foreach my $val (@val) {
        if ((defined ($val)) && ($val !~ m/^[\s\t\r\n]*$/)) {
            push (@VAL, $val);
        }
    }
    if (!@VAL) {
        return;
    }
    $id = $self->pused ('id', $id);
    $code = $self->pused ('code', $code);
    if (!exists ($self->{'data'}{$type}{$id})) {
        $self->{'data'}{$type}{$id} = {};
    }
    if (!exists ($self->{'data'}{$type}{$id}{$code})) {
        $self->{'data'}{$type}{$id}{$code} = {};
    }
    foreach my $val (@VAL) {
        if ($val =~ m/^[0-9]+$/) {
            $self->{'data'}{$type}{$id}{$code}{$val} = 1;
        } else {
            $val =~ s/\\/\\\\/g;
            $val =~ s/"/\\"/g;
            $self->{'data'}{$type}{$id}{$code}{'"' . $val . '"'} = 1;
        }
    }
}

sub pused
{
    my ($self, $type, $code) = @_;

    if ($code =~ m/^https?:/) {
        return ('<' . $code . '>');
    }
    if ($code =~ m/^([-0-9A-Za-z]+):/) {
        my $pf = $1;
        if (!exists ($self->{'prefix'}{$pf})) {
            die ("fatal: unknown prefix '$pf' in $type: $code");
        }
        $self->{'pused'}{$pf} = 1;
        return ($code);
    } else {
        if ((($type ne 'code') || ($code ne 'a')) && ($code !~ m/^(true|false)$/)) {
            die ("fatal: missing prefix in $type: $code");
        }
        return ($code);
    }
}

sub output
{
    my ($self, $basefile) = @_;

    my $len = 0;
    foreach my $key (sort (keys (%{$self->{'pused'}}))) {
        if (length ($key) > $len) {
            $len = length ($key);
        }
    }
    $len++;
    my $prefix = '';
    foreach my $key (sort (keys (%{$self->{'pused'}}))) {
        $prefix .= sprintf ("\@prefix %-${len}s <%s> .\n", $key . ':', $self->{'prefix'}{$key});
    }
    $prefix .= "\n";
    my $fou;
    if ($basefile eq '-') {
        $fou = *STDOUT;
        binmode ($fou, ':utf8');
        print ($fou $prefix);
    }
    foreach my $type (sort (keys (%{$self->{'data'}}))) {
        my $fc = '01';
        my $lc = 60;
        if ($basefile ne '-') {
            if (!open ($fou, "> $basefile-$type-$fc.ttl")) {
                die ("fatal: failed to open '$basefile-$type-$fc.ttl' for output: $!");
            }
            binmode ($fou, ':utf8');
            print ($fou $prefix);
        }
        foreach my $id (sort (keys (%{$self->{'data'}{$type}}))) {
            printf ($fou "%s\n", $id);
            $lc++;
            my @codes = sort (keys (%{$self->{'data'}{$type}{$id}}));
            my $code;
            while ($code = shift (@codes)) {
                my $pos = '.';
                if (@codes) {
                    $pos = ';';
                }
                if (length ($code) > 30) {
                    printf ($fou "    %s\n", $code);
                    printf ($fou "    %-30s %s %s\n", '', join (' , ', sort (keys (%{$self->{'data'}{$type}{$id}{$code}}))), $pos);
                    $lc += 2;
                } else {
                    printf ($fou "    %-30s %s %s\n", $code, join (' , ', sort (keys (%{$self->{'data'}{$type}{$id}{$code}}))), $pos);
                    $lc++;
                }
            }
            if ($lc >= 10000000) {
                $fc = sprintf ('%02d', ($fc + 1));
                close ($fou);
                if (!open ($fou, "> $basefile-$type-$fc.ttl")) {
                     die ("fatal: failed to open '$basefile-$type-$fc.ttl' for output: $!");
                }
                binmode ($fou, ':utf8');
                print ($fou $prefix);
                $lc = 60;
            }
        }
        if ($basefile ne '-') {
            close ($fou);
        }
    }
}

sub output_nt
{
    my ($self, $basefile) = @_;

    my $fou;
    if ($basefile eq '-') {
        $fou = *STDOUT;
        binmode ($fou, ':utf8');
    }
    foreach my $type (sort (keys (%{$self->{'data'}}))) {
        if ($basefile ne '-') {
            if (!open ($fou, "> $basefile-$type.nt")) {
                die ("fatal: failed to open '$basefile-$type.nt' for output: $!");
            }
            binmode ($fou, ':utf8');
        }
        foreach my $id (sort (keys (%{$self->{'data'}{$type}}))) {
            my $nt1 = $self->output_nt_value ($id);
            foreach my $code (sort (keys (%{$self->{'data'}{$type}{$id}}))) {
                my $nt2 = $self->output_nt_value ($code);
                foreach my $val (sort (keys (%{$self->{'data'}{$type}{$id}{$code}}))) {
                    print ($fou join (' ', $nt1, $nt2, $self->output_nt_value ($val)), " .\n");
                }
            }
        }
        if ($basefile ne '-') {
            close ($fou);
        }
    }
}

sub output_nt_value
{
    my ($self, $val) = @_;

    if ($val =~ s/^([a-z]+)://) {
        if ($self->{'prefix'}{$1}) {
            return ('<' . $self->{'prefix'}{$1} . $val . '>');
        } else {
            die ("fatal: unknown prefix '$1' in: $1:$val\n");
        }
    }
    if ($val =~ m/^[<"]/) {
        return ($val);
    }
    if ($val eq 'a') {
        return ('<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>');
    }
    if ($val =~ m/^[0-9]+$/) {
        return ('"' . $val . '"^^<http://www.w3.org/2001/XMLSchema#integer>');
    }
    if ($val =~ m/^(true|false)$/) {
        return ('"' . $val . '"^^<http://www.w3.org/2001/XMLSchema#boolean>');
    }
    die ("fatal: unexpected value: $val\n");
}

sub prefix_code
{
    my ($self, $uri) = @_;

    $uri =~ s/^[<\s]+//;
    $uri =~ s/[>\s]+$//;
    foreach my $pre (keys (%{$self->{'prefix'}})) {
        if ($self->{'prefix'}{$pre} eq $uri) {
            return ($pre);
        }
    }
    return (undef);
}

1;

